### PR TITLE
[FSDP][1/N] Rework `clip_grad_norm_()` and tests

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #87480 [FSDP][2/N] Remove `params_with_grad`
* **#87479 [FSDP][1/N] Rework `clip_grad_norm_()` and tests**
* #87478 [FSDP][Docs] Clarify warnings to mention collectives

This PR reworks FSDP's `clip_grad_norm_()` and its unit tests. The unit tests in `test_fsdp_core.py` still need to be revisited and will be done in follow-up work.

Some details in arbitrary order:
- This renames `_calc_grad_norm()` to `_get_grad_norm()`. This is to simplify our verb usage in method names. Otherwise, we may diverge to different verbs like "compute", "calculate", "get", "find" etc. I am open to discussion here.
- Because we call `torch.linalg.vector_norm()` as the underlying norm calculation subroutine, which can take infinity as input for the norm type, there is no reason to have a separate conditional branch for the infinity norm.
- This removes a host-device synchronization point from `clip_grad_norm_()` by using the same trick from `torch.nn.utils.clip_grad_norm_()`. This may improve throughput for workloads like metaseq, which computes gradient norms regularly.
- This returns the total norm from `clip_grad_norm_()` as mentioned in the docstring. Before nothing was returned.
- This rewrites the unit tests, which were slightly problematic. Much of the logic to verify gradient norms were computed correctly were exactly the same as the logic used to compute them in FSDP (i.e. `^p`, sum via all-reduce, `^(1/p)`). This defeats the purpose of unit testing. There were some other oddities like `input = torch.rand(14, 2, device=self.rank); in_data = torch.tensor(input[self.rank], device=self.rank)`, where we materialize a full `(14, 2)` shape but only ever use the first two rows (assuming world size 2).

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 1]
```diff
@@ -39,67 +41,133 @@
...
-        "cpu_offload",
-        [CPUOffload(offload_params=True), CPUOffload(offload_params=False)],
-    )
-    def test_fsdp_clip_grad_norm(self, norm_type, nested_fsdp, cpu_offload):
-        """Test FSDP with clip grad norm."""
-        self._run_fsdp_one_iteration(norm_type, nested_fsdp, cpu_offload)
+        # Multiply gradients by a large factor to ensure that gradients will
+        # actually be clipped
+        for param in itertools.chain(ddp_model.parameters(), fsdp_model.parameters()):
+            if param.grad is not None:  # gradients may be `None` for `use_orig_params=True`
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:135]: why would some grads be none, is it possible not all params are used in loss computation?

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 2]
```diff
@@ -39,67 +41,133 @@
...
+
+        ddp_total_norm = torch.nn.utils.clip_grad_norm_(
+            ddp_model.parameters(), max_norm=max_norm, norm_type=norm_type,
+        )
+        fsdp_total_norm = fsdp_model.clip_grad_norm_(max_norm=max_norm, norm_type=norm_type)
+        self.assertEqual(ddp_total_norm, fsdp_total_norm)
 
+        # Check that the gradients were modified by `clip_grad_norm_()`
+        for param, orig_grad in zip(ddp_model.parameters(), orig_ddp_grads):
+            assert not torch.equal(param.grad, orig_grad)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:151]: nit: instead of `assert`, use `self.assertNotEqual`?

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 3]
```diff
@@ -39,67 +41,133 @@
...
+        for param, orig_grad in zip(fsdp_model.parameters(), orig_fsdp_grads):
+            if param.grad is None:
+                self.assertEqual(param.grad, orig_grad)  # `None`
+            else:
+                assert not torch.equal(param.grad, orig_grad)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:156]: ditto

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 4]
```diff
@@ -39,67 +41,133 @@
...
-        self.assertEqual(total_norm, total_norm_expected)
+        # Run an optimizer step to ensure gradients matched after clipping
+        ddp_optim.step()
+        fsdp_optim.step()
+        with FSDP.summon_full_params(fsdp_model):
+            for (n1, p1), (n2, p2) in zip(
+                ddp_model.module.named_parameters(),
+                fsdp_model.named_parameters(),
+            ):
+                self.assertEqual(n1, n2)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:166]: nit: checking names is nice but probably beyond the scope of this unittest?

**Review by rohan-varma — APPROVED** [review]
LGTM, some minor comments inline

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py (hunk 2 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:151]: I use `assert` when it is part of the test infrastructure itself. If this `assert` fails, then the unit test is meaningless; it does not indicate that the unit test failed.

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:135]: Parameters not in the local shard have `None` gradient.

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py (hunk 4 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:166]: I agree.

To be honest, I tend to do this while testing so that `n1` and `n2` are not unused variables. It is often helpful when the subsequent parameter equality check fails to see the name of the violating parameter. The issue is that if you override the error string to add the name, you lose the normal error string. I wonder if there is a way to simply _append_ a custom string to the normal error string.
